exec/run.go

```diff
@@ -14,38 +18,103 @@ type Cmd struct {
 	Name    string   // Executable name.
 	Argv    []string // Executable arguments.
 	Command string   // Shell command.
+	Dir     string   // The Command's working directory.
 
-	Dir string // The Command's working directory.
+	Timeout time.Duration // Specifies the amount of time a command is allowed to run.
+	Retries int           // Amount of times a command can be retried.
 
 	// If neither Env nor WithEnv are set, the environment is inherited from os.Environ().
 	Env     map[string]string // If set, the command's environment is _set_ to Env.
 	WithEnv map[string]string // If set, the command's environment is _added_ to WithEnv.
 }
 
 // Run executes a `Cmd`.
-func Run(cmd Cmd) (stdout, stderr string, err error) {
+func Run(cmd Cmd) (string, string, error) {
+	var stdout, stderr string
+	var err error
+
+	for i := 0; i <= cmd.Retries; i++ {
+		if cmd.Timeout != 0 {
+			stdout, stderr, err = runWithTimeout(cmd)
+		} else {
+			log.WithFields(log.Fields{
+				"name": cmd.Name,
+				"argv": cmd.Argv,
+			}).Debug("called Run")
+
+			xc, stderrBuf := BuildExec(cmd)
+
+			log.WithFields(log.Fields{
+				"dir": xc.Dir,
+				"env": xc.Env,
+			}).Debug("executing command")
+
+			var stdoutBuf []byte
+			stdoutBuf, err = xc.Output()
+			stdout = string(stdoutBuf)
+			stderr = stderrBuf.String()
+
+			log.WithFields(log.Fields{
+				"stdout": stdout,
+				"stderr": stderr,
+			}).Debug("done running")
+		}
+
+		if err == nil {
+			break
+		}
+	}
+
+	return stdout, stderr, err
+}
+
+// RunTimeout executes a `Cmd` and waits to see if it times out.
+func runWithTimeout(cmd Cmd) (string, string, error) {
+	seconds := cmd.Timeout
 	log.WithFields(log.Fields{
 		"name": cmd.Name,
 		"argv": cmd.Argv,
-	}).Debug("called Run")
+	}).Debug("called Start")
 
-	xc, stderrBuf := BuildExec(cmd)
+	xc, stderr := BuildExec(cmd)
+	var stdout strings.Builder
+	xc.Stdout = &stdout
+
+	err := xc.Start()
+	if err != nil {
+		return "", "", errors.Wrap(err, "Error starting command ")
+	}
 
 	log.WithFields(log.Fields{
 		"dir": xc.Dir,
 		"env": xc.Env,
 	}).Debug("executing command")
 
-	stdoutBuf, err := xc.Output()
-	stdout = string(stdoutBuf)
-	stderr = stderrBuf.String()
-
-	log.WithFields(log.Fields{
-		"stdout": stdout,
-		"stderr": stderr,
-	}).Debug("done running")
+	done := make(chan error, 1)
+	go func() {
+		done <- xc.Wait()
+	}()
+
+	select {
+	case <-time.After(seconds):
+		err := xc.Process.Kill()
+		if err != nil {
+			return "", "", errors.Wrapf(err, "Error killing the process")
+		}
+
+		return "", "", errors.New(fmt.Sprintf("Operation timed out running `%s %s` after %s", cmd.Name, cmd.Argv, cmd.Timeout))
```

Comment:
@dchenk I think this is something I corrected you on a while back. I now realize this is due to the inconsistency of the error package we are using in different packages. I think consolidating this and creating a standard is a task for another PR.